Fix API reprocessing image when saving in EditContent.vue

MrDirkelz opened the pull request.

MrDirkelz committed:
Add a change that will prevent the image from being reprocessed in the API from EditContent.vue

ivanslabbert commented on cms/src/components/content/EditContent.vue:
```diff
@@ -240,6 +240,21 @@ const isDirty = computed(
+    saveBtnText.value = "Saving";
+    // Use a timeout to wait for the API to process the new data, and then refresh the parent document
+    // This prevents things like the image from reprocessing in the API
+    setTimeout(() => {
```
Using a timer here will be unpredictable - we do not know how long it will take to get an API response

A better approach could be to

1. Use a useDexieLiveQuery to get live updates of the parent document from IndexedDB
2. When the image field updates in the database version of the parent document, patch the `editableParent` with the update image data

MrDirkelz committed:
Refactor BasePage template structure and improve loading state handling in EditContent

MrDirkelz committed:
Fix conditional rendering in EditContent to ensure editableParent is defined

MrDirkelz committed:
Refactor EditContent to simplify dirty checking and improve image data handling

MrDirkelz committed:
Enhance dirty checking in EditContent by refining data loading and change detection logic

MrDirkelz committed:
Show network changes popup to user and disable save button so the user does pay attention so that the previous changes that were made won't be lost and takes preference

MrDirkelz committed:
Remove unused template debug printed networkChanges variable

MrDirkelz committed:
Add a check to see if imageData does exist in editableParent to prevent undefined type error from occuring

MrDirkelz committed:
Fix the tests that were failing in EditContent.spec

MrDirkelz committed:
Refactor code in EditContent.spec && EditContent.vue to potentially fix type-errors

MrDirkelz committed:
Rather delete properties that reduce data accuracy in dirty checking than trying to omit them, as it causes type error warnings

MrDirkelz committed:
Only delete _rev property if it does exist

The _rev property may or may not exist, as it is mostly just a database
field in couchDB I think

MrDirkelz committed:
Cast content as any in dirty checking for Network Changes in EditContent.spec

The _.rev field exists at runtime, but not in our DTOs, to prevent type
errors we can just cast content temporarily as any, as we know _rev will
exist at runtime.

MrDirkelz committed:
Changes made to how the copy of content for network changes dirty checking

As updatedTimeUtc is not optional I can't delete it temporarily for
dirty checking without casting content as any temporarily for this
special circumstance

DirkDevelopmentFreelance commented:
Change mock express api port to be a random port to change it each time

ivanslabbert commented on cms/src/components/content/EditContent.spec.ts:
```diff
@@ -256,36 +259,50 @@ describe("EditContent.vue", () => {
+                expect(wrapper.find('button[data-test="save-button"]').exists()).toBe(true); // EditContentParentValidation is rendered
+            });
+        },
+        { timeout: 999999 },
```
remove timeout

ivanslabbert reviewed (commented):
Can you add a test to make sure the logic is working as expected? (Create new ticket?)

MrDirkelz committed:
Remove timeout from test in EditContent.spec

MrDirkelz committed:
Remove network changes functionality\

The network changes functionality introduced bugs unrelated to the main
focus of this ticket. It was decided to create a new ticket for this
functionality. So this ticket can be merged successfully.

MrDirkelz committed:
Remove left over changes for network changes functionality

MrDirkelz committed:
Remove unneccesary disabled attribute on the save button in EditContent

MrDirkelz committed:
Refactor isDirty to be more understandable and less buggy

MrDirkelz committed:
Fix naming of a new function inside isDirty in EditContent

MrDirkelz committed:
Implicity type parameters in isDirty helper function to fix type errors

MrDirkelz committed:
Changed updatedBy to updatedTimeUtc in omitFields function inside isDirty checking

MrDirkelz committed:
Add updatedBy field to omit in the omitFields function for parent documents

MrDirkelz committed:
Omit relevant fields for dirty checking in EditContentParentValidation component

MrDirkelz committed:
Fix dirty checking for images in the patch image watcher

MrDirkelz committed:
Ommit the correct fields in EditContentValidation to prevent buggy behavior

DirkDevelopmentFreelance commented:
Bad PR
